### PR TITLE
Revert to compiling PHP from source

### PR DESCRIPTION
**Pull request by mnapoli**
Because of #42 and https://github.com/remicollet/remirepo/issues/228

Fixes https://github.com/brefphp/aws-lambda-layers/issues/42

TODO left:

- [x] Build `bref/fpm-internal-src`
- [x] Cleanup lib-check
- [x] Adapt the release scripts
- [x] Update README

### CODE REVIEW

File: tests/test_2_extensions.php [hunk 1]
```diff
@@ -79,6 +79,8 @@
     'json' => function_exists('json_encode'),
     'bcmath' => function_exists('bcadd'),
     'ctype' => function_exists('ctype_digit'),
+    // https://github.com/brefphp/aws-lambda-layers/issues/42
+    'curl-with-http2' => defined('CURL_HTTP_VERSION_2'),
```
Comment (deleugpn) [inline tests/test_2_extensions.php:83]: NICE!

**Comment by ciaranmcnulty** [comment]
I like the use of `bake` here so far, looking good 👍 

I think the zip should still happen inside the builds, you can easily dump to disk using `output` in HCL targets

**Comment by mnapoli** [comment]
@ciaranmcnulty Do you mean buildx output flags like this: https://docs.docker.com/engine/reference/commandline/buildx_build/#output ? I don't see how we can export a single directory as a zip with that, but I'm interested.

I removed the zip stage in Docker in order to simplify the scripts, and because building (and caching) more docker images on GitHub Actions is already such a mess. I know it breaks the full "docker encapsulation", but it really makes things simpler. I'm open to alternatives though!

**Comment by ciaranmcnulty** [comment]
Yes, I mean something like the `zip-fpm` stages but combined into one:

```Dockerfile
FROM alpine:3.14 AS zip
RUN apk add zip

# ... maybe do this for all of the versions? the named context populated in bake.hcl
COPY --from=fpm-80 /opt /opt
RUN zip --quiet --recurse-paths /tmp/layer.zip .

FROM scratch AS output

COPY --from=zip /tmp/layer.zip /
```

You could then build the `output` target into a dir

**Comment by ciaranmcnulty** [comment]
Does the base-level image have a use outside of being effectively a build cache? Is there a reason it's shipped as an image rather than pushing the cache layers to a registry?

Maybe out of scope for this PR but the caching approach could have a bit of an overhaul with use of `--no-cache-filter` and some stage refactoring to make specific parts of the build cache-bustable but enable the builds to resume later



```

**Comment by ciaranmcnulty** [comment]
There's significant overlap between the Dockerfiles of different architectures, having everything split out makes this a bit tricky to see what's going on.

I'm not going to argue it's unambiguously easier to read but take a look at how we structured things in https://github.com/jakzal/phpqa/blob/master/Dockerfile and use `FROM --platform=` to ensure some parts of the build run on the builder's native platform

**Comment by mnapoli** [comment]
> Does the base-level image have a use outside of being effectively a build cache? Is there a reason it's shipped as an image rather than pushing the cache layers to a registry?

Indeed. Unless I'm missing something, it's now no longer pushed anywhere, just built for the build process. (it used to be published to Docker Hub)

> Maybe out of scope for this PR but the caching approach could have a bit of an overhaul with use of --no-cache-filter and some stage refactoring to make specific parts of the build cache-bustable but enable the builds to resume later

I had a look at that option, but that would only cache _less_ things so I'm not sure how that would improve things here?

> There's significant overlap between the Dockerfiles of different architectures, having everything split out makes this a bit tricky to see what's going on.

Good point, thanks! The base-devel images had no reason to be split up anymore. If we indeed need to have different builds in the future, we can split them up again.

But I pushed a commit where I merged the 2 files together, so that's a win!

Now I need to figure out why the ARM builds have started failing with the cryptic errors:

> buildx bake failed with: ERROR: failed to solve: failed to copy: invalid status response 403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature. for https://vth0acprodeus2file0.blob.core.windows.net/daf72de42ddf4a22bb30b289c092b345/edc875780696ed11bf7a14cb652f465e?sv=2019-07-07&sr=b&sig=H1DVEU9wjTKeu03F8GMeV5vgP37AF9joAj4IF505jMA%3D&se=2023-01-21T12%3A32%3A57Z&sp=r&rscl=x-e2eid-f69ae48d-47ec417e-9d28c90a-d03bf09a

Maybe the build is taking too long and the credentials for uploading stuff to GitHub Actions cache have expired… Maybe there are too many requests. Maybe it's something else 🤷

**Comment by mnapoli** [comment]
> Good point, thanks! The base-devel images had no reason to be split up anymore. If we indeed need to have different builds in the future, we can split them up again.
>
> But I pushed a commit where I merged the 2 files together, so that's a win!

🤦 that was actually also the case for all the other Dockerfiles. I merged all of them together and reduced the number of lines of code drastically, thank you!

image

**Comment by deleugpn** [comment]
I know this won't be super relevant, but I just wanted to point out why I chose two different files. Variables makes things harder to mentally parse. You need to know possible values upfront and beginners unaware of how the system works are much less likely to know/understand what these variables will be.

At the time when we were not compiling PHP anymore, I felt that it was a lot more useful to have static/duplicated code than code that has variables spread around it and having to understand the many places these variables flow around. In the end, we've spent 30+ years with a single CPU architecture and today we have 2. Chances are that we won't need more than 2 for another decade or so. Variables bring more value when they help reuse the code significantly more than just 2 times. 

Part of this is moot point with compiling though, as it appears the entire Dockerfile is practically the same. It does make the `Makefile` a bit more complex to reason about, but everything is about trade-offs.

**Comment by mnapoli** [comment]
It's all working now, builds are fast (see #44), I'm very happy with the results and:

image

we switched to a more complex build while not exploding in complexity 🎉

Thanks a lot @ciaranmcnulty for the advice, and @deleugpn for adding context (indeed we're iterating a lot on the structure as the context drastically changes every time, the previous build had completely different Dockerfiles for different architectures).

Anyway, I gave the new layers a try and it's all working, tests are passing, let's get this merged and benchmark it!

**Pull request merged**